### PR TITLE
Service instance profile configuration

### LINKED ISSUE #256: How can I connect to multiple vCenter from the same minion?
I have added the configuration as follows on the master

top.sls
```
base:
  m1:
    - vc1
    - vc2
```

v1.sls
```
vmware_config:
  host: 192.168.100.10
  password: Password@123
  user: salt-test@vsphere.local
  verify_ssl: false
```

v2.sls
```
vmware_config:
  host: 192.168.100.20
  password: Password@123
  user: salt-test@vsphere.local
  verify_ssl: false
```

Now, I cannot seem to figure out how to call the commands using the specified credentials. For example, query datacenter list from `vc2`.

### PR DESCRIPTION
**Pull request by abhi1693**
This is a draft for adding support for `profiles` for service instances. The changes should not break any existing functionality. However, there could be more changes needed here. Since I'm unaware of most of the code, I stayed away from doing too many changes that may impact the existing code base.

I have removed the `get_username_password` method as it seemed to be used in only 2 places and replaced it with `get_config` that should handle data being passed from either `__opts__`, `__pillar__` or `env`. It also looks for the `esxi_host` name if it is passed to the method.

I ran the test suite using `VMware-vCenter-Server-Appliance-7.0.3.00500-19480866`

I'm not particularly sure about the files under the `states` module. I have test cases failing for `esxi` and `license_mgr` in there and I cannot seem to figure out why. It could be my changes to `conftest.py` are incorrect.

Also, I don't have a `vmc` setup and all test cases failed related to that failed on my system. This is out of scope for me as I have no knowledge of this product.

Fixes #256 

cc: @waynew

### CODE REVIEW

File: src/saltext/vmware/utils/connect.py [hunk 1]
```diff
@@ -78,26 +77,38 @@ def get_service_instance(opts=None, pillar=None, esxi_host=None):
                     user: admin
                     password: ***
 
+    If configuration for multiple VMWare services instances is required, they can be
+    set up as different configuration profiles:
+    For example:
+    .. code-block:: yaml
+
+        vmware_config:
```
Comment (waynew) [inline src/saltext/vmware/utils/connect.py:85]: **suggestion** this would become `saltext.vmware:`

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -84,24 +84,39 @@ def service_instance(integration_test_config):
...
     setattr(cluster_ha_mod, "__pillar__", {})
+    setattr(cluster_ha_mod, "__salt__", vmware_conf)
     setattr(cluster_drs_mod, "__opts__", {})
     setattr(cluster_drs_mod, "__pillar__", {})
+    setattr(cluster_drs_mod, "__salt__", vmware_conf)
     setattr(esxi_mod, "__pillar__", vmware_conf)
     setattr(esxi_mod, "__opts__", {})
+    setattr(esxi_mod, "__salt__", vmware_conf)
     setattr(esxi_st, "__pillar__", vmware_conf)
+    setattr(esxi_st, "__salt__", vmware_conf)
```
Comment (waynew) [inline tests/integration/conftest.py:106]: **suggestion** it's not necessarily the responsibility for this function, but this fixture is entirely wrong. Either we should access these via the salt loader (make calls via the salt_call_cli), or  we should use the `configure_loader_modules` fixture, which ensures that faked attributes like this are cleaned up.

File: src/saltext/vmware/utils/connect.py [hunk 3]
```diff
@@ -78,26 +77,38 @@ def get_service_instance(opts=None, pillar=None, esxi_host=None):
...
+
+    config = get_config(config=config, profile=profile, esxi_host=esxi_host)
+
     service_instance = connect.SmartConnect(
-        host=config["host"],
-        user=config["user"],
-        pwd=config["password"],
+        host=config[0],
+        user=config[1],
+        pwd=config[2],
```
Comment (waynew) [inline src/saltext/vmware/utils/connect.py:111]: **question** can we not use key access here? Index access is easier to break without things really breaking :thinking:

**Review by waynew — CHANGES_REQUESTED** [review]
Definitely loving this! I didn't see anything that made me go :eyes: - mostly just suggestions about how to improve some things even further, since we're actually making some pretty significant improvements here.

I don't have any particularly strong feelings about the setattr fixture being modified/fixed as part of this PR, but we should definitely lock in the naming scheme. 

I've created [a discussion about the config scheme](https://github.com/saltstack/salt-ext-modules-vmware/discussions/282) if you want to weigh in over there :+1:

File: tests/integration/conftest.py (hunk 2 above)
Comment (abhi1693) [reply tests/integration/conftest.py:106]: I'm not sure about that at all or even how to fix it. I just followed the convention I saw here.

File: src/saltext/vmware/utils/connect.py [hunk 4]
```diff
@@ -23,40 +23,39 @@
...
+def get_config(config, profile=None, esxi_host=None):
+    if profile:
+        credentials = config.get("vmware_config", {})[profile]
+    else:
+        credentials = config.get("vmware_config", {})
+
+    if esxi_host:
+        host = os.environ.get("VMWARE_CONFIG_HOST") or credentials.get("esxi_host", {}).get("host")
+    else:
+        host = os.environ.get("VMWARE_CONFIG_HOST") or credentials.get("host")
```
[the remark this answers is not in the document]
Comment (abhi1693) [reply src/saltext/vmware/utils/connect.py]: I have made the changes but I tried to be very cautious so as to not make too many breaking ones. Let me know if you need anything else changed in here.

File: src/saltext/vmware/utils/connect.py (hunk 1 above)
Comment (abhi1693) [reply src/saltext/vmware/utils/connect.py]: Done

File: src/saltext/vmware/utils/connect.py (hunk 3 above)
Comment (abhi1693) [reply src/saltext/vmware/utils/connect.py]: Done

**Comment by abhi1693** [comment]
@waynew The PR is ready for another round of review. Once everything looks good, I'll also resolve the conflicts.

**Comment by abhi1693** [comment]
@waynew I have a few more changes I would like to contribute to the repo which I'm using in my organization, however, those greatly depend on this PR being merged into the main repo. I have also opened a few new issues stating the changes needed.

I'd only need help with the test cases, and I can keep contributing more changes every ~2 weeks as we are heavily dependent on VMWare deployments and we need this repo to provide all methods needed to make our automation work flawlessly.

**Comment by waynew** [comment]
I'm currently OOO, but I'll review this when I get back

**Comment by waynew** [comment]
I'm currently OOO, but I'll review this when I get back - should be the 30th or the 1st

**Comment by waynew** [comment]
Getting to take a look at this today - I also just opened a PR against your fork/branch to fix the docs build (which will probably fail here on CI in a few minutes).

Just resolved all the outstanding comments - looks like there was a pre-commit failure that also needs to be addressed.

**Comment by abhi1693** [comment]
@waynew I have made the changes for `pre-commit` issues.

Also, you are a day early :)

**Comment by abhi1693** [comment]
@waynew Do we have any ETA on this being added to the main branch? Because I'd like to get discussion rolling for other things I've been doing for the repo.

**Comment by waynew** [comment]
We've got the 4th holiday on Monday, and I still need to (re)set up my integration test infra. It may be Tuesday that I get to (re)review/test/merge this but probably more like Wednesday.

**Comment by abhi1693** [comment]
Please see https://github.com/saltstack/salt-ext-modules-vmware/discussions/282#discussioncomment-3069787 when you get a chance. I made some changes to my old branch regarding the configuration and it now works much better. Check: https://github.com/Onemind-Services-LLC/salt-ext-modules-vmware/commit/e99da6d6376c954167a07f84e2f6271a7e3cdf7c#diff-c0dea8f0e5a8ea3b0b1af9e1374080e9afe2a3740665c790d47b1a3ed77a4971R26

Happy holidays :)

**Comment by abhi1693** [comment]
@waynew did you get a chance to look into this?

**Comment by abhi1693** [comment]
@waynew Hey, I hate to bug you so much but I was hoping to move this forward so that we can work on several other ideas and fixes I have pending for this repo.

**Comment by abhi1693** [comment]
Can we discuss to move this forward? @waynew

**Comment by waynew** [comment]
@abhi1693 absolutely! Sorry it's been a while - I was OOO for VMware Explore and it's just been hectic in general around here. I'm settling things down a bit now so I should be able to get back to this and your other PR. I'm really liking what I'm seeing so far after refreshing a little bit. @joechainz and @chruck should also take a look at this PR and weigh in.

**Comment by waynew** [comment]
Okay, apologies once again for the wait - had a high priority item take over, but I'm looking at this *right now* :tada:

**Comment by abhi1693** [comment]
> Okay, apologies once again for the wait - had a high priority item take over, but I'm looking at this *right now* :tada: 

Thanks. Before you merge please make sure you review this as well https://github.com/saltstack/salt-ext-modules-vmware/pull/258#issuecomment-1172930931

File: tests/integration/conftest.py (hunk 2 above)
Comment (waynew) [reply tests/integration/conftest.py:106]: Oooo... now I see. Yeah, we can remove `__salt__` from here for sure.

**Review by waynew — CHANGES_REQUESTED** [review]
Okay, so this I think looks *mostly* good. I went ahead and checked this out and hacked on it a bit - the `__salt__` -> `__opts__` thing definitely needs to be updated. Other than that, I wasn't able to connect to my own vSphere with this, but it might have been a config issue locally - I'll give it a shot again tomorrow and see if I can figure out what was happening... I was getting this message:

```
Passed invalid arguments: Required field "userName" not provided (not @optional).   
```

File: src/saltext/vmware/utils/connect.py [hunk 5]
```diff
@@ -23,83 +23,90 @@
...
+
     if config["host"] is None or config["password"] is None or config["user"] is None:
         raise ValueError("Cannot create service instance, VMware credentials incomplete.")
+
     service_instance = connect.SmartConnect(
-        host=config["host"],
-        user=config["user"],
-        pwd=config["password"],
+        host=config.get("host"),
+        user=config.get("username"),
```
Comment (waynew) [inline src/saltext/vmware/utils/connect.py:108]: **suggestion** this should also be changed to `user` - see the return from `def get_config`

**Review by waynew — CHANGES_REQUESTED** [review]

File: src/saltext/vmware/utils/connect.py (hunk 5 above)
Comment (abhi1693) [reply src/saltext/vmware/utils/connect.py:108]: That's strange that I missed that. I'll fix that in a few minutes

Comment (abhi1693) [reply src/saltext/vmware/utils/connect.py]: Done

**Review by cheburakshu — APPROVED** [review]

File: src/saltext/vmware/modules/cluster.py [hunk 6]
```diff
@@ -39,7 +39,7 @@ def list_(service_instance=None):
     """
     ret = {}
     if service_instance is None:
-        service_instance = get_service_instance(opts=__opts__, pillar=__pillar__)
+        service_instance = get_service_instance(config=__salt__, profile=profile)
```
[the remark this answers is not in the document]
Comment (waynew) [reply src/saltext/vmware/modules/cluster.py]: I guess it depends on your shell :joy:

**Review by waynew — APPROVED** [review]
I think this looks good now - there are a couple of things that do need some fixin', but they should have been protected by some tests, so I'll go ahead and make those changes in another PR :+1:

**Pull request merged**